controllers/machine.go

```diff
@@ -82,45 +93,92 @@ func (mrc *machineReconcileContext) addFinalizer() error {
 	return nil
 }
 
-func (mrc *machineReconcileContext) ensureDependencies() error {
-	hardware, err := mrc.ensureHardware()
-	if err != nil {
-		return fmt.Errorf("ensuring hardware: %w", err)
-	}
+func (mrc *machineReconcileContext) isHardwareReady(hw *tinkv1.Hardware) bool {
+	return hw.Spec.Metadata.State == inUse && hw.Spec.Metadata.Instance.State == provisioned
+}
+
+type errRequeueRequested struct{}
+
+func (e *errRequeueRequested) Error() string {
+	return "requeue requested"
+}
+
+func (mrc *machineReconcileContext) ensureTemplateAndWorkflow(hw *tinkv1.Hardware) (*tinkv1.Workflow, error) {
+	wf, err := mrc.workflow()
+
+	switch {
+	case apierrors.IsNotFound(err):
+		if err := mrc.ensureTemplate(hw); err != nil {
+			return nil, fmt.Errorf("failed to ensure template: %w", err)
+		}
+
+		if err := mrc.createWorkflow(hw); err != nil {
+			return nil, fmt.Errorf("failed to create workflow: %w", err)
+		}
 
-	if err := mrc.ensureTemplate(hardware); err != nil {
-		return fmt.Errorf("ensuring template: %w", err)
+		return nil, &errRequeueRequested{}
+	case err != nil:
+		return nil, fmt.Errorf("failed to get workflow: %w", err)
+	default:
 	}
 
-	if err := mrc.ensureWorkflow(hardware); err != nil {
-		return fmt.Errorf("ensuring workflow: %w", err)
+	s := wf.GetCurrentActionState()
+	if s == tinkv1.WorkflowStateFailed || s == tinkv1.WorkflowStateTimeout {
+		mrc.log.Info("current action is in a failed or timed out state: %v", wf.GetCurrentAction())
```

Comment:
Should we not return with an Error here? Or are we going to proceed to mark machine ready even if workflow has failed?

Reply:
i had this error out but Chris suggested this option.

Reply:
this will just log a message but yeah proceed on.

Reply:
But i think we should not proceed to mark ready when a workflow has failed?
This would log and still return the workflow, and if it was the last workflow that has failed it would mark the machine as ready.

In case an earlier action on the workflow had failed, then yes! Machine wont go to ready.

Reply:
i'm ok to update this and have the reconcile func return error. @chrisdoherty4 thoughts?